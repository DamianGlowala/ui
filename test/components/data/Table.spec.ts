import { describe, it, expect, vi } from 'vitest'
import { UTable } from '#components'
import type { TypeOf } from 'zod'
import ComponentRender from '../component-render'

describe('Table', () => {
  it.each([
    [
      'basic case',
      {
        props: {
          columns: [{ key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }],
          ui: { tr: { base: '' }, th: { base: '' }, emptyState: { wrapper: 'flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14' } }
        }
      }
    ],
    [
      'sort',
      {
        props: {
          sort: { column: 'name', direction: 'desc' }, columns: [{ key: 'name', label: 'Name', sortable: true }, { key: 'age', label: 'Age' }],
          ui: { tr: { base: '' }, th: { base: '' }, emptyState: { wrapper: 'flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14' }, default: { sortButton: { icon: 'i-heroicons-academic-cap-16-solid' } } }
        }
      }
    ],
    [
      'sort mode',
      {
        props: {
          sortMode: 'manual', columns: [{ key: 'name', label: 'Name', sortable: true }, { key: 'age', label: 'Age' }],
          ui: { tr: { base: '' }, th: { base: '' }, emptyState: { wrapper: 'flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14' }, default: { sortButton: { icon: 'i-heroicons-academic-cap-16-solid' } } }
        }
      }
    ],
    [
      'custom sort button',
      {
        props:
        {
          sortButton: { label: 'Custom button', size: 'xl', variant: 'orange', icon: 'i-heroicons-academic-cap' }, columns: [{ sortable: true, key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }],
          ui: { tr: { base: '' }, th: { base: '' }, emptyState: { wrapper: 'flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14' }, default: { sortButton: { icon: 'i-heroicons-academic-cap' } } }
        }
      }
    ],
    [
      'sort asc icon',
      {
        props:
        {
          sortAscIcon: 'i-heroicons-adjustments-horizontal', sort: { column: 'name', direction: 'asc' }, columns: [{ sortable: true, key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }],
          ui: { tr: { base: '' }, th: { base: '' }, td: { base: '' }, emptyState: { wrapper: 'flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14' }, default: { sortButton: { icon: 'i-heroicons-academic-cap' } } }
        }
      }
    ],
    [
      'sort desc icon',
      {
        props: {
          sortDescIcon: 'i-heroicons-arrow-down-left-solid', sort: { column: 'name', direction: 'desc' }, columns: [{ sortable: true, key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }],
          ui: { tr: { base: '' }, th: { base: '' }, td: { base: '' }, emptyState: { wrapper: 'flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14' }, default: { sortButton: { icon: 'i-heroicons-academic-cap' } } }
        }
      }
    ],
    [
      'loading state',
      {
        props:
        {
          loading: true, loadingState: { icon: 'i-heroicons-sparkles', label: 'Loading...' }, columns: [{ key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }],
          ui: { tr: { base: '' }, th: { base: '' }, loadingState: { wrapper: 'flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14' } }
        }
      }
    ],
    [
      'loading state slot',
      {
        slots: { 'loading-state': 'Loading state slot' },
        props:
        {
          loading: true, loadingState: { icon: 'i-heroicons-sparkles', label: 'Loading...' }, columns: [{ key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }],
          ui: { tr: { base: '' }, th: { base: '' }, loadingState: { wrapper: 'flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14' } }
        }
      }
    ],
    ['empty state props',
      {
        props:
        {
          rows: [], emptyState: { icon: 'i-heroicons-inbox', label: 'No data available' }, columns: [{ key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }],
          ui: { tr: { base: '' }, th: { base: '' }, emptyState: { wrapper: 'flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14' } }
        }
      }
    ],
    ['empty state slot',
      {
        slots: { 'empty-state': 'Empty state slot' },
        props: {
          emptyState: { label: 'emptyStateLabel', icon: 'i-heroicons-chevron-up-down-solid' }, columns: [{ key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }],
          ui: { tr: { base: '' }, th: { base: '' }, emptyState: { wrapper: 'flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14' } }
        }
      }
    ],
    ['selected rows',
      {
        props: {
          rows: [{ id: 1, name: 'John' }, { id: 2, name: 'Jane' }, { id: 3, name: 'Doe' }], columns: [{ key: 'name', label: 'Name' }, { key: 'age', label: 'Age' }],
          ui: { tr: { base: '' }, th: { base: '' }, td: { base: '' }, emptyState: { wrapper: 'flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14' } }
        }
      }
    ],

    // @ts-ignore
  ])('renders %s correctly', async (nameOrHtml: string, options: TypeOf<typeof Table.props>) => {
    if (options !== undefined) {
      options.slots = options.slots || { default: () => 'label' }
      options.slots.default = options.slots.default || (() => 'label')
    }
    const html = await ComponentRender(nameOrHtml, options, UTable)
    expect(html).toMatchSnapshot()
  })
})
