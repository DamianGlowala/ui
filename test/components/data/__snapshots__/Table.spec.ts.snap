// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > renders basic case correctly 1`] = `
"<div>
  <table class="">
    <thead class="">
      <tr class="">
        <!--v-if-->
        <th scope="col" class=""><span>Name</span></th>
        <th scope="col" class=""><span>Age</span></th>
      </tr>
    </thead>
    <tbody class="">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-circle-stack-20-solid" aria-hidden="true"></span>
            <p class="">No items.</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders columns attribute correctly 1`] = `
"<div>
  <table class="">
    <thead class="">
      <tr class="">
        <!--v-if-->
        <th scope="col" class=""><span></span></th>
        <th scope="col" class=""><span></span></th>
      </tr>
    </thead>
    <tbody class="">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-circle-stack-20-solid" aria-hidden="true"></span>
            <p class="">No items.</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders custom sort button correctly 1`] = `
"<div>
  <table class="">
    <thead class="">
      <tr class="">
        <!--v-if-->
        <th scope="col" class=""><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-base gap-x-2.5 px-3.5 py-2.5 inline-flex items-center"><span class="i-heroicons-academic-cap flex-shrink-0 h-6 w-6" aria-hidden="true"></span><span class="">Name</span>
            <!--v-if-->
          </button></th>
        <th scope="col" class=""><span>Age</span></th>
      </tr>
    </thead>
    <tbody class="">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-circle-stack-20-solid" aria-hidden="true"></span>
            <p class="">No items.</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders empty state props correctly 1`] = `
"<div>
  <table class="">
    <thead class="">
      <tr class="">
        <!--v-if-->
        <th scope="col" class=""><span>Name</span></th>
        <th scope="col" class=""><span>Age</span></th>
      </tr>
    </thead>
    <tbody class="">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-inbox" aria-hidden="true"></span>
            <p class="">No data available</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders empty state slot correctly 1`] = `
"<div>
  <table class="">
    <thead class="">
      <tr class="">
        <!--v-if-->
        <th scope="col" class=""><span>Name</span></th>
        <th scope="col" class=""><span>Age</span></th>
      </tr>
    </thead>
    <tbody class="">
      <tr>
        <td colspan="2">Empty state slot</td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders loading state correctly 1`] = `
"<div>
  <table class="">
    <thead class="">
      <tr class="">
        <!--v-if-->
        <th scope="col" class=""><span>Name</span></th>
        <th scope="col" class=""><span>Age</span></th>
      </tr>
    </thead>
    <tbody class="">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-sparkles" aria-hidden="true"></span>
            <p class="">Loading...</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders loading state slot correctly 1`] = `
"<div>
  <table class="">
    <thead class="">
      <tr class="">
        <!--v-if-->
        <th scope="col" class=""><span>Name</span></th>
        <th scope="col" class=""><span>Age</span></th>
      </tr>
    </thead>
    <tbody class="">
      <tr>
        <td colspan="2">Loading state slot</td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders selected rows correctly 1`] = `
"<div>
  <table class="">
    <thead class="">
      <tr class="">
        <!--v-if-->
        <th scope="col" class=""><span>Name</span></th>
        <th scope="col" class=""><span>Age</span></th>
      </tr>
    </thead>
    <tbody class="">
      <tr class="">
        <!--v-if-->
        <td class="">John</td>
        <td class=""></td>
      </tr>
      <tr class="">
        <!--v-if-->
        <td class="">Jane</td>
        <td class=""></td>
      </tr>
      <tr class="">
        <!--v-if-->
        <td class="">Doe</td>
        <td class=""></td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders sort asc icon correctly 1`] = `
"<div>
  <table class="">
    <thead class="">
      <tr class="">
        <!--v-if-->
        <th scope="col" class=""><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
            <!--v-if--><span class="">Name</span><span class="i-heroicons-adjustments-horizontal flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
          </button></th>
        <th scope="col" class=""><span>Age</span></th>
      </tr>
    </thead>
    <tbody class="">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-circle-stack-20-solid" aria-hidden="true"></span>
            <p class="">No items.</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders sort correctly 1`] = `
"<div>
  <table class="">
    <thead class="">
      <tr class="">
        <!--v-if-->
        <th scope="col" class=""><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
            <!--v-if--><span class="">Name</span><span class="i-heroicons-bars-arrow-down-20-solid flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
          </button></th>
        <th scope="col" class=""><span>Age</span></th>
      </tr>
    </thead>
    <tbody class="">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-circle-stack-20-solid" aria-hidden="true"></span>
            <p class="">No items.</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders sort desc icon correctly 1`] = `
"<div>
  <table class="">
    <thead class="">
      <tr class="">
        <!--v-if-->
        <th scope="col" class=""><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
            <!--v-if--><span class="">Name</span><span class="i-heroicons-arrow-down-left-solid flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
          </button></th>
        <th scope="col" class=""><span>Age</span></th>
      </tr>
    </thead>
    <tbody class="">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-circle-stack-20-solid" aria-hidden="true"></span>
            <p class="">No items.</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`Table > renders sort mode correctly 1`] = `
"<div>
  <table class="">
    <thead class="">
      <tr class="">
        <!--v-if-->
        <th scope="col" class=""><button type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center -m-1.5">
            <!--v-if--><span class="">Name</span><span class="i-heroicons-arrows-up-down-20-solid flex-shrink-0 h-5 w-5" aria-hidden="true"></span>
          </button></th>
        <th scope="col" class=""><span>Age</span></th>
      </tr>
    </thead>
    <tbody class="">
      <tr>
        <td colspan="2">
          <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14"><span class="i-heroicons-circle-stack-20-solid" aria-hidden="true"></span>
            <p class="">No items.</p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;
